File: app/src/main/java/ru/visualmath/android/api/model/QuestionBlockSlide.java
package ru.visualmath.android.api.model;

import com.google.gson.Gson;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;

import org.json.JSONException;
import org.json.JSONObject;

import java.lang.reflect.Type;

public class QuestionBlockSlide {

    private QuestionBlock questionBlock;

    private int index;

    private boolean started;

    public QuestionBlock getQuestionBlock() {
        return questionBlock;
    }

    public int getIndex() {
        return index;
    }

    public boolean isStarted() {
        return started;
    }

    public static class Deserializer implements JsonDeserializer<QuestionBlockSlide> {

        @Override
        public QuestionBlockSlide deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) {
            QuestionBlockSlide slide = new QuestionBlockSlide();

            try {
                JSONObject jsonObject = new JSONObject(json.toString());

                String contentJson = jsonObject.getJSONObject("content").toString();
                slide.questionBlock = new Gson().fromJson(contentJson, QuestionBlock.class);
                slide.started = !jsonObject.isNull("activeContent");
                if (slide.started) {
                    JSONObject activeContent = jsonObject.getJSONObject("activeContent");
                    slide.started = !activeContent.getBoolean("ended");
                    String currentQuestion = activeContent.optString("currentQuestion");
                    if ("done".equals(currentQuestion)) {
                        slide.index = -1;
                    } else {
                        slide.index = new Gson().fromJson(currentQuestion, Integer.class);
                    }
                }
            } catch (JSONException ignored) {
            }
            return slide;
        }
    }
}

File: app/src/main/java/ru/visualmath/android/api/model/QuestionBlockSlide.kt
package ru.visualmath.android.api.model

import com.google.gson.Gson
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.annotations.SerializedName

import org.json.JSONException
import org.json.JSONObject

import java.lang.reflect.Type

data class QuestionBlockSlide(

        @SerializedName("questionBlock")
        var questionBlock: QuestionBlock,

        @SerializedName("index")
        var index: Int,

        @SerializedName("started")
        var isStarted: Boolean
) {
    class Deserializer : JsonDeserializer<QuestionBlockSlide?> {

        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): QuestionBlockSlide? {
            var slide: QuestionBlockSlide? = null

            try {
                val jsonObject = JSONObject(json.toString())

                val contentJson = jsonObject.getJSONObject("content").toString()
                val questionBlock = Gson().fromJson(contentJson, QuestionBlock::class.java)
                var index = 0
                var isStarted = !jsonObject.isNull("activeContent")
                if (isStarted) {
                    val activeContent = jsonObject.getJSONObject("activeContent")
                    isStarted = !activeContent.getBoolean("ended")
                    val currentQuestion = activeContent.optString("currentQuestion")
                    index = if ("done" == currentQuestion) -1 else Gson().fromJson(currentQuestion, Int::class.java)
                }
                slide = QuestionBlockSlide(questionBlock, index, isStarted)
            } catch (ignored: JSONException) {
            }

            return slide
        }
    }
}
